Applications/FileIO/GocadIO/GocadTSurfaceReader.cpp

```diff
@@ -0,0 +1,525 @@
+        if (line.empty() || isCommentLine(line))
+            continue;
+        else if (str[0] == "GOCAD_ORIGINAL_COORDINATE_SYSTEM")
+        {
+            Gocad::CoordinateSystem coordinate_system;
+            if (!coordinate_system.parse(in))
+            {
+                ERR("Error parsing coordinate system.");
+                return false;
+            }
+        }
+        else if (str[0] == "GEOLOGICAL_FEATURE" ||
+                 str[0] == "GEOLOGICAL_TYPE" ||
+                 str[0] == "STRATIGRAPHIC_POSITION")
+        {
+            // geological and stratigraphic information - currently ignored
+        }
+        else if (str[0] == "PROPERTY_CLASS_HEADER")
+        {
+            if (!parsePropertyClass(in))
+            {
+                ERR("Error parsing PROPERTY_CLASS_HEADER.");
+                return false;
+            }
+        }
+        else if (str[0] == "PROPERTIES")
+        {
+            if (!parseProperties(in, str, mesh_prop))
+            {
+                ERR("Error parsing PROPERTIES");
+                return false;
+            }
+        }
+        else if (str[0] == "TFACE")
+        {
+            std::vector<MeshLib::Node*> nodes;
+            std::vector<MeshLib::Element*> elems;
+            std::map<std::size_t, std::size_t> node_id_map;
+            INFO ("Parsing surface %s", mesh_name.c_str());
+            if (!parseSurface(in, nodes, elems, node_id_map, mesh_prop))
+            {
+                ERR("Error parsing Surface %s.", mesh_name.c_str());
+                clearData(nodes, elems);
+                return false;
+            }
+            _mesh_vec.push_back(
+                new MeshLib::Mesh(mesh_name, nodes, elems, mesh_prop));
+            return true;
+        }
+        else
+        {
+            WARN("GocadTSurfaceReader::readMesh() - Unknown keyword found: %s",
+                 line.c_str());
+        }
+    }
+    ERR("%s", eof_error.c_str());
+    return false;
+}
+
+MeshLib::Mesh* GocadTSurfaceReader::getData(std::size_t const idx) const
+{
+    if (_mesh_vec.empty())
+    {
+        ERR("Error: No mesh data available.");
+        return nullptr;
+    }
+    if (idx < _mesh_vec.size())
+        return _mesh_vec[idx];
+    ERR("Error: Mesh index (%d) out of range (0, %d).", idx, _mesh_vec.size());
+    return nullptr;
+}
+
+std::vector<MeshLib::Mesh*> GocadTSurfaceReader::getData() const
+{
+    if (_mesh_vec.empty())
+    {
+        ERR("Error: No mesh data available.");
+    }
+    return _mesh_vec;
+}
+
+std::string GocadTSurfaceReader::getMeshName(std::size_t idx) const
+{
+    if (_mesh_vec.empty())
+    {
+        ERR("Error: No mesh data available.");
+        return std::string();
+    }
+    if (idx < _mesh_vec.size())
+        return _mesh_vec[idx]->getName();
+    ERR("Error: Mesh index (%d) out of range (0, %d).", idx, _mesh_vec.size());
+    return std::string();
+}
+
+void GocadTSurfaceReader::writeData(std::string const& file_name,
```

Comment:
Why does the reader need the possibility to write the meshes. This could be done also by external functionality, for  instance the DataExplorer.

Reply:
The mesh writing functionality including all checks and path construction currently covers >40 LoC that otherwise would need to be duplicated in both DE and CLI. In my opinion it makes sense, that an interface offers functionality to save data with a single call if it outputs something more complex than a single mesh or something comparable. In this case it's probably not obvious which processing steps have to be performed to savely get and write the data (i.e. one mesh vs multiple meshes, path construction on windows vs linux, etc.). Other file interfaces do the same thing.

If you prefer that I move that functionality to another file (TSurfaceIO or whatever) that is certainly possible. But I won't include (and duplicate) all that code in the DE, that is not the place for IO functionality!

Reply:
:heavy_check_mark: